sympy/solvers/diophantine/diophantine.py

```diff
@@ -51,6 +52,108 @@
     "univariate"}
 
 
+class DiophantineSolutionSet(set):
+    """
+    Container for a set of solutions to a particular diophantine equation.
+
+    The base representation is a set of tuples representing each of the solutions.
+
+    Parameters
+    ==========
+
+    symbols : list
+        List of free symbols in the original equation.
+    parameters: list (optional)
+        List of parameters to be used in the solution.
+
+    Examples
+    ========
+
+    Adding solutions:
+
+        >>> from sympy.solvers.diophantine.diophantine import DiophantineSolutionSet
+        >>> from sympy.abc import x, y, t, u
+        >>> s1 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s1
+        set()
+        >>> s1.add((2, 3))
+        >>> s1.add((-1, u))
+        >>> s1
+        {(-1, u), (2, 3)}
+        >>> s2 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s2.add((3, 4))
+        >>> s1.update(*s2)
+        >>> s1
+        {(-1, u), (2, 3), (3, 4)}
+
+    Conversion of solutions into dicts:
+
+        >>> list(s1.dict_iterator())
+        [{x: -1, y: u}, {x: 2, y: 3}, {x: 3, y: 4}]
+
+    Substituting values:
+
+        >>> s3 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s3.add((t**2, t + u))
+        >>> s3
+        {(t**2, t + u)}
+        >>> s3.subs(2, 3)
+        {(4, 5)}
+        >>> s3.subs(-1)
+        {(1, u - 1)}
+        >>> s3.subs(t=3)
+        {(9, u + 3)}
+        >>> s3(-2, 3)
+        {(4, 1)}
+        >>> s3(5)
+        {(25, u + 5)}
+    """
+
+    def __init__(self, symbols_list, parameters=None):
+        super().__init__()
+        self.symbols = symbols_list
+
+        if parameters is None:
+            self.parameters = symbols('%s1:%i' % ('t', len(self.symbols) + 1), integer=True)
+        else:
+            self.parameters = parameters
+
+        self.kw_parameter_mapping = {str(p): p for p in self.parameters}
+
+    def add(self, solution):
+        if len(solution) != len(self.symbols):
+            raise ValueError("Solution is of incorrect length")
+        super(DiophantineSolutionSet, self).add(Tuple(*solution))
+
+    def update(self, *solutions):
+        for solution in solutions:
+            self.add(solution)
+
+    def dict_iterator(self):
+        for solution in sorted(self, key=lambda s: tuple(s)):
+            yield dict(zip(self.symbols, solution))
+
+    def subs(self, *values, **kw_values):
+        if len(values) > len(self.parameters):
+            raise ValueError("Substitution has too many values")
+
+        result = DiophantineSolutionSet(self.symbols, self.parameters)
+
+        subs_dict = dict(zip(self.parameters, values))
+
+        for parameter in kw_values:
+            if parameter not in self.kw_parameter_mapping:
```

Comment:
What happens if they pass `**{t: 1}`? 
```python
>>> x in ['x']
False
```

Reply:
SymPy already matches by name so I don't think the `kw_parameter_mapping` is necessary:
```python
>>> var('x',real=True)
x
>>> x.subs('x',1)
1
```

Reply:
That example does not work the way you say in my environment:

```
>>> from sympy import *
>>> var('x',real=True)
x
>>> x.subs('x',1)
x
```

Reply:
hmm...if I do `del x` before example I get the same. So it looks like this does not always work and a mapping would be necessary.